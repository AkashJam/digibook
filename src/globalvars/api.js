const path = "http://10.0.2.2:3000";

const API = {
  // user
  login: async (user) => {
    //POST json
    //making data to send on server
    // const data = { username: username, password: password };
    //POST request
    const response = await fetch(`${path}/user/login`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(user),
    })
      .then((response) => response.json())
      //If response is in json then in success
      .then((responseJson) => {
        return responseJson;
      })
      //If response is not in json then in error
      .catch((error) => {
        return error;
      });
    return response;
  },
  register: async (user) => {
    //POST json
    //making data to send on server
    // const data = { username: username, password: password };
    //POST request
    const response = await fetch(`${path}/user/register`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(user),
    })
      .then((response) => response.json())
      //If response is in json then in success
      .then((responseJson) => {
        return responseJson;
      })
      //If response is not in json then in error
      .catch((error) => {
        return error;
      });
    return response;
  },
  updateUser: async (userdata) => {
    //POST json
    //making data to send on server
    // const data = { username: username, password: password };
    //POST request
    const response = await fetch(`${path}/user/login`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(userdata),
    })
      .then((response) => response.json())
      //If response is in json then in success
      .then((responseJson) => {
        return responseJson;
      })
      //If response is not in json then in error
      .catch((error) => {
        return error;
      });
    return response;
  },

  // groups
  getGroups: async (userID) => {
    //GET request
    const response = await fetch(`${path}/group/${userID}`, {
      method: "GET",
      //Request Type
    })
      .then((response) => response.json())
      //If response is in json then in success
      .then((responseJson) => {
        //Success
        return responseJson;
      })
      //If response is not in json then in error
      .catch((error) => {
        //Error
        return error;
      });
    return response;
  },
  createGroup: async (groupdata) => {
    //POST json
    //making data to send on server
    // const data = { username: username, password: password };
    //POST request
    const response = await fetch(`${path}/group/add`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(groupdata),
    })
      .then((response) => response.json())
      //If response is in json then in success
      .then((responseJson) => {
        return responseJson;
      })
      //If response is not in json then in error
      .catch((error) => {
        return error;
      });
    return response;
  },
  updateGroup: async (groupdata) => {
    //POST json
    //making data to send on server
    // const data = { username: username, password: password };
    //POST request
    const response = await fetch(`${path}/group/update`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(groupdata),
    })
      .then((response) => response.json())
      //If response is in json then in success
      .then((responseJson) => {
        return responseJson;
      })
      //If response is not in json then in error
      .catch((error) => {
        return error;
      });
    return response;
  },

  // Tasks
  getTasks: async (userID) => {
    //GET request
    const response = await fetch(`${path}/task/${userID}`, {
      method: "GET",
      //Request Type
    })
      .then((response) => response.json())
      //If response is in json then in success
      .then((responseJson) => {
        //Success
        return responseJson;
      })
      //If response is not in json then in error
      .catch((error) => {
        //Error
        return error;
      });
    return response;
  },
  createTask: async (taskdata) => {
    //POST json
    //making data to send on server
    // const data = { username: username, password: password };
    //POST request
    const response = await fetch(`${path}/task/add`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(taskdata),
    })
      .then((response) => response.json())
      //If response is in json then in success
      .then((responseJson) => {
        return responseJson;
      })
      //If response is not in json then in error
      .catch((error) => {
        return error;
      });
    return response;
  },
  updateTask: async (taskdata) => {
    //POST json
    //making data to send on server
    // const data = { username: username, password: password };
    //POST request
    const response = await fetch(`${path}/task/update`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(taskdata),
    })
      .then((response) => response.json())
      //If response is in json then in success
      .then((responseJson) => {
        return responseJson;
      })
      //If response is not in json then in error
      .catch((error) => {
        return error;
      });
    return response;
  },
};

export default API;
