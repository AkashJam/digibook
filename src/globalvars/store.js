import React, { useReducer } from "react";
import { toastr } from "./";
import AsyncStorage from "@react-native-async-storage/async-storage";

const user = {
  id: 0,
  username: "",
  displayName: null,
  range: 0.75,
  collaborators: [],
  groups: [],
  activities: [],
  notifications: [],
  locations: {},
};

const storeUserLog = (data) =>
  AsyncStorage.setItem(`DN_userlog`, JSON.stringify(data));

const reducer = (state, action) => {
  let log = [];
  let obj = {};
  switch (action.type) {
    //users

    case "set_user":
      // console.log(action.locations);
      obj = {
        id: action.user.id,
        username: action.user.username,
        displayName: action.user.displayName,
        range: action.user.range,
        collaborators: action.user.collaborators,
        groups: action.user.groups,
        activities: action.user.activities,
        notifications: action.notifications,
        locations: action.locations,
      };
      storeUserLog(obj);
      toastr(
        `Welcome ${
          action.user.displayName !== null
            ? action.user.displayName
            : action.user.username
        }`
      );
      return obj;

    case "update_display_name":
      // obj = await API.updateUser({
      //   id: state.id,
      //   task: { display_name: action.displayName },
      // });
      // if (obj.code == 200) {
      obj = {
        id: state.id,
        username: state.username,
        displayName: action.displayName,
        range: state.range,
        collaborators: action.collaborators,
        groups: state.groups,
        activities: state.activities,
        notifications: state.notifications,
        locations: state.locations,
      };
      storeUserLog(obj);
      toastr("Display name updated.");
      return obj;
    // } else {
    //   toastr("Error connecting to server.", toastTime);
    //   return state;
    // }

    case "update_user_range":
      // obj = await API.updateUser({
      //   id: state.id,
      //   user: { range: action.range },
      // });
      // if (obj.code == 200) {
      obj = {
        id: state.id,
        username: state.username,
        displayName: state.displayName,
        range: action.range,
        collaborators: state.collaborators,
        groups: state.groups,
        activities: state.activities,
        notifications: state.notifications,
        locations: state.locations,
      };
      storeUserLog(obj);
      toastr("Notification range updated.");
      return obj;
    // } else {
    //   toastr("Error connecting to server.", toastTime);
    //   return state;
    // }

    //activities
    case "add_task":
      const newTask = {
        id: action.id,
        description: action.description,
        group_id: action.group_id,
        notify: false,
        location: null,
        datetime: action.datetime !== undefined ? action.datetime : null,
        completed: false,
        createdOn: new Date().toString(),
        modifiedOn: null,
      };
      log = [newTask, ...state.activities];
      obj = { ...state, activities: log };
      storeUserLog(obj);
      toastr("Task added.");
      return obj;

    case "rename_task":
      log = state.activities.map((activity) => {
        if (activity.id === action.id) {
          return {
            ...activity,
            description: action.description,
            modifiedOn: new Date().toString(),
          };
        } else {
          return activity;
        }
      });
      obj = { ...state, activities: log };
      storeUserLog(obj);
      toastr("Task renamed.");
      return obj;

    case "delete_task":
      log = state.activities.filter((activity) => activity.id !== action.id);
      obj = { ...state, activities: log };
      storeUserLog(obj);
      toastr("Task deleted.");
      return obj;

    case "change_group":
      log = state.activities.map((activity) => {
        if (activity.id === action.id) {
          return {
            ...activity,
            group_id: action.group_id,
            modifiedOn: new Date().toString(),
          };
        } else {
          return activity;
        }
      });
      obj = { ...state, activities: log };
      storeUserLog(obj);
      toastr("Group changed.");
      return obj;

    case "change_datetime":
      log = state.activities.map((activity) => {
        if (activity.id === action.id) {
          return {
            ...activity,
            datetime: action.datetime,
            modifiedOn: new Date().toString(),
          };
        } else {
          return activity;
        }
      });
      obj = { ...state, activities: log };
      storeUserLog(obj);
      toastr("Date and time changed.");
      return obj;

    case "change_location":
      log = state.activities.map((activity) => {
        if (activity.id === action.id) {
          return {
            ...activity,
            location: action.location,
            modifiedOn: new Date().toString(),
          };
        } else {
          return activity;
        }
      });
      obj = { ...state, activities: log };
      storeUserLog(obj);
      toastr("Location changed.");
      return obj;

    case "toggle_notification":
      log = state.activities.map((activity) => {
        if (activity.id === action.id) {
          return {
            ...activity,
            notify: action.notify,
            modifiedOn: new Date().toString(),
          };
        } else {
          return activity;
        }
      });
      obj = { ...state, activities: log };
      storeUserLog(obj);
      toastr(action.notify ? "Notification On." : "Notification Off.");
      return obj;

    case "toggle_completion":
      log = state.activities.map((activity) => {
        if (activity.id === action.id) {
          return {
            ...activity,
            completed: action.completed,
            datetime: new Date().toString(),
            modifiedOn: new Date().toString(),
          };
        } else {
          return activity;
        }
      });
      obj = { ...state, activities: log };
      storeUserLog(obj);
      toastr(
        action.completed ? "Task completed." : "Task no longer completed."
      );
      return obj;

    // groups
    case "add_group":
      let newGroup = {
        id: action.id,
        name: action.name,
        createdOn: new Date().toString(),
        modifiedOn: null,
      };
      log = [newGroup, ...state.groups];
      obj = { ...state, groups: log };
      storeUserLog(obj);
      toastr("Group added.");
      return obj;

    case "rename_group":
      // obj = await API.updateGroup({
      //   id: action.id,
      //   group: { name: action.name },
      // });
      // if (obj.code == 200) {
      log = state.groups.map((group) => {
        if (group.id === action.id) {
          return {
            ...group,
            name: action.name,
            modifiedOn: new Date().toString(),
          };
        } else {
          return group;
        }
      });
      obj = { ...state, groups: log };
      storeUserLog(obj);
      toastr("Group renamed.");
      return obj;
    // } else {
    //   toastr("Error connecting to server.", toastTime);
    //   return state;
    // }

    case "delete_group":
      // obj = await API.updateGroup({
      //   id: action.id,
      //   group: { active: false },
      // });
      // if (obj.code == 200) {
      log = state.groups.filter((group) => group.id !== action.id);
      obj = { ...state, groups: log };
      storeUserLog(obj);
      toastr("Group deleted.");
      return obj;
    // } else {
    //   toastr("Error connecting to server.", toastTime);
    //   return state;
    // }

    //notifications
    case "update_notifications":
      // obj = await API.updateUser({
      //   id: state.id,
      //   task: { display_name: action.displayName },
      // });
      // if (obj.code == 200) {
      obj = {
        id: state.id,
        username: state.username,
        displayName: state.displayName,
        range: state.range,
        collaborators: state.collaborators,
        groups: state.groups,
        activities: state.activities,
        notifications: action.notifications,
        locations: state.locations,
      };
      // storeUserLog(obj);
      // toastr("Display name updated.");
      return obj;
    // } else {
    //   toastr("Error connecting to server.", toastTime);
    //   return state;
    // }

    // locations
    case "set_location":
      obj = state.locations ? state.locations : {};
      if (
        !state.locations ||
        state.locations.lat === null ||
        state.locations.lon === null
      ) {
        obj.lat = action.latitude;
        obj.lon = action.longitude;
      } else if (
        Math.abs(obj.latitude - action.latitude) > 0.01 &&
        Math.abs(obj.longitude - action.longitude) > 0.01
      ) {
        Object.keys(obj).forEach((e) => {
          if (e !== "lat" && e !== "lon") obj[e] = [];
        });
        obj.lat = action.latitude;
        obj.lon = action.longitude;
      }
      obj[`${action.use}`] = action.location;
      storeUserLog({ ...state, locations: obj });
      return { ...state, locations: obj };

    default:
      return state;
  }
};

const UserContext = React.createContext({
  state: user,
  dispatch: () => null,
});

const UserProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, user);

  return (
    <UserContext.Provider value={[state, dispatch]}>
      {children}
    </UserContext.Provider>
  );
};

export { UserContext, UserProvider };
