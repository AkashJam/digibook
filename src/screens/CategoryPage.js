import React, { useState, useEffect } from "react";
import {
  View,
  FlatList,
  Modal,
  StyleSheet,
  Pressable,
  Text,
  Alert,
} from "react-native";
import { UserContext, toastr, API } from "../globalvars";
import { COLORS, FONTS, PAGE, SIZES } from "../constants";
import DropDownPicker from "react-native-dropdown-picker";
import {
  GestureHandlerRootView,
  TextInput,
  TouchableOpacity,
} from "react-native-gesture-handler";
import { Card, CollabMenu, ControlButton, Header } from "../components";
import { useNavigation } from "@react-navigation/native";
import { Octicons, Ionicons, FontAwesome } from "@expo/vector-icons";

export default function CategoryPage() {
  const navigation = useNavigation();
  const [state, dispatch] = React.useContext(UserContext);
  const [open, setOpen] = useState(false);
  const def = state.groups.filter((group) => group.name === "All");
  const [value, setValue] = useState(def[0].id);
  let categories = [];
  state.groups.forEach((element) => {
    categories.push({ label: element.name, value: element.id });
  });
  const [items, setItems] = useState(categories);
  const [edit, setEdit] = useState(false);
  const [name, setName] = useState("");
  const [collabMenu, setCollabMenu] = useState(false);
  const [users, setUsers] = useState([]);
  const [owner, setOwner] = useState(false);
  const [groupMenu, setGroupMenu] = useState(false);

  let data = {};

  useEffect(() => {
    if (items.length > state.groups.length) {
      console.log(state.groups.length);
      manageGroup.addGroup(items[items.length - 1].label);
    }
  }, [items]);

  useEffect(async () => {
    if (value !== def[0].id) {
      const userdata = await API.getUserGroups(value);
      setUsers(userdata);
      console.log(users)
      const groupinfo = userdata.filter(
        (user) => user.username === state.username
      );
      if (groupinfo[0].owner) setOwner(true);
      else setOwner(false);
    } else setOwner(true);
  }, [value]);

  const manageTask = {
    addTask: async (taskName) => {
      try {
        data = await API.createTask({
          id: value,
          task: {
            description: taskName,
          },
        });
        console.log(data);
        if (data.code == 200)
          dispatch({
            type: "add_task",
            id: data.id,
            description: taskName,
            group_id: value,
          });
        else toastr(data.status);
      } catch (error) {
        console.log(error);
      }
    },
    setCompleted: async (id, value) => {
      try {
        data = await API.updateTask({
          id: id,
          task: { completed: value, datetime: new Date() },
        });
        if (data.code == 200)
          dispatch({ type: "toggle_completion", id: id, completed: value });
        else toastr(data.status);
      } catch (error) {
        console.log(error);
      }
    },
    notifyToggle: async (id, value) => {
      try {
        data = await API.updateTask({
          id: id,
          task: { notify: value },
        });
        if (data.code == 200)
          dispatch({ type: "toggle_notification", id: id, notify: value });
        else toastr(data.status);
      } catch (error) {
        console.log(error);
      }
    },
    deleteTask: (id) => {
      Alert.alert("Delete Task", "Are you sure you want to delete this task?", [
        {
          text: "Cancel",
          style: "cancel",
        },
        {
          text: "Yes",
          onPress: async () => {
            data = await API.updateTask({
              id: id,
              task: { active: false },
            });
            if (data.code == 200) dispatch({ type: "delete_task", id: id });
            else toastr(data.status);
          },
        },
      ]);
    },
  };

  const manageGroup = {
    addGroup: async (groupName) => {
      try {
        data = await API.createGroup({
          id: state.id,
          group: { name: groupName },
        });
        console.log(data);
        if (data.code == 200) {
          dispatch({
            type: "add_group",
            name: groupName,
            id: data.id,
          });
          let updatedList = items;
          updatedList[updatedList.length - 1].value = data.id;
          setItems(updatedList);
          setValue(data.id);
        } else toastr(data.status);
      } catch (error) {
        console.log(error);
      }
    },
    renameGroupToggle: () => {
      const group = state.groups.filter((group) => group.id === value);
      setName(group[0].name);
      setEdit(true);
      setGroupMenu(false);
    },
    renameGroup: async () => {
      if (name !== "") {
        setEdit(false);
        const group = state.groups.filter((group) => group.id === value);
        if (name !== group[0].name) {
          data = await API.updateGroup({
            id: value,
            group: { name: name },
          });
          console.log(data);
          if (data.code == 200) {
            const newItems = items.map((item) => {
              if (item.value === value) return { label: name, value: value };
              else return item;
            });
            setItems(newItems);
            dispatch({ type: "rename_group", id: value, name: name });
          } else toastr(data.status);
        }
      } else toastr("Group needs a name");
    },
    toggleCollabMenu: async () => {
      // const data = await API.getUserGroups(value);
      // setUsers(data);
      setGroupMenu(false);
      setCollabMenu(true);
    },
    deleteGroup: () => {
      setGroupMenu(false);
      Alert.alert(
        "Delete Group",
        "Are you sure you want to delete group?\nAll tasks in group will be removed as well",
        [
          {
            text: "Cancel",
            style: "cancel",
          },
          {
            text: "Yes",
            onPress: async () => {
              if (owner)
                data = await API.updateGroup({
                  id: value,
                  group: { active: false },
                });
              else
                data = await API.updateUserGroup({
                  id: value,
                  user: { username: state.username },
                  group: { active: false },
                });
              if (data.code == 200) {
                dispatch({ type: "delete_group", id: value });
                const newItems = items.filter((item) => item.value !== value);
                setItems(newItems);
                setValue(def[0].id);
              } else toastr(data.status);
            },
          },
        ]
      );
    },
  };

  const editGroupMenu = () => {
    if (edit) setEdit(false);
    else setGroupMenu(!groupMenu);
  };

  const Menu = () => {
    if (value === def[0].id) return <></>;
    return (
      <View
        style={{
          ...styles.menuContainer,
          padding: groupMenu ? SIZES.margin : 0,
          borderColor: groupMenu ? COLORS.accent : COLORS.primary,
          backgroundColor: groupMenu ? COLORS.secondary : COLORS.primary,
        }}
      >
        <TouchableOpacity style={{ marginLeft: 20 }} onPress={editGroupMenu}>
          {!groupMenu && !edit && (
            <Octicons name="kebab-horizontal" size={32} color={COLORS.accent} />
          )}
          {groupMenu && (
            <Ionicons name="close" size={32} color={COLORS.accent} />
          )}
          {edit && <Ionicons name="close" size={32} color={COLORS.accent} />}
        </TouchableOpacity>
        {groupMenu && (
          <View style={styles.options}>
            <TouchableOpacity onPress={manageGroup.renameGroupToggle}>
              <Text style={styles.menuText}>Rename</Text>
            </TouchableOpacity>
            <TouchableOpacity onPress={manageGroup.toggleCollabMenu}>
              <Text style={styles.menuTextAlt}>Manage</Text>
            </TouchableOpacity>
            <TouchableOpacity onPress={manageGroup.deleteGroup}>
              <Text style={styles.menuText}>Delete</Text>
            </TouchableOpacity>
          </View>
        )}
      </View>
    );
  };

  return (
    <>
      <Header screen={"Categories"} />
      <Menu />
      <CollabMenu
        active={collabMenu}
        close={() => setCollabMenu(false)}
        id={value}
        owner={owner}
        users={users}
        setUsers={(newUsers) => setUsers(newUsers)}
      />
      <View style={PAGE}>
        {edit && (
          <View style={styles.group}>
            <TouchableOpacity
              style={styles.button}
              onPress={manageGroup.renameGroup}
            >
              <FontAwesome name="edit" color={COLORS.accent} size={32} />
            </TouchableOpacity>
            <TextInput
              style={styles.textInput}
              value={name}
              onChangeText={(text) => setName(text)}
              onBlur={() => setEdit(false)}
            />
          </View>
        )}
        {!edit && (
          <DropDownPicker
            style={styles.dropdown}
            dropDownContainerStyle={styles.container}
            textStyle={{
              color: COLORS.secondary,
              ...FONTS.h1_bold,
            }}
            listItemLabelStyle={FONTS.p_regular}
            searchTextInputStyle={FONTS.p_regular}
            searchable={true}
            searchPlaceholder="Enter a new name to create group"
            addCustomItem={true}
            open={open}
            value={value}
            items={items}
            setOpen={setOpen}
            setValue={setValue}
            setItems={setItems}
          />
        )}
        <GestureHandlerRootView style={{ flex: 1, marginVertical: 20 }}>
          <FlatList
            data={state.activities.filter(
              (task) => value === def[0].id || task.group_id === value
            )}
            renderItem={({ item, index }) => (
              <Card
                task={item}
                index={index}
                edit={false}
                setEdit={(id) => navigation.navigate("Activity", { id })}
                {...manageTask}
              />
            )}
            keyExtractor={(item) => `${item.id}`}
          />
        </GestureHandlerRootView>
        <ControlButton addTask={manageTask.addTask} />
      </View>
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: COLORS.accent,
    marginHorizontal: SIZES.margin,
    width: "96%",
    padding: 10,
    borderWidth: 0,
  },
  dropdown: {
    backgroundColor: COLORS.accent,
    paddingHorizontal: 20,
    margin: SIZES.margin,
    width: "96%",
    borderRadius: SIZES.borderRadius,
    borderWidth: 0,
  },
  menuContainer: {
    borderWidth: 2,
    marginRight: "1%",
    position: "absolute",
    top: 45,
    right: SIZES.padding,
    zIndex: 4,
    borderRadius: SIZES.borderRadius,
    flexDirection: "row-reverse",
  },
  options: {
    // marginTop: SIZES.padding,
    // marginRight: SIZES.padding,
    // paddingRight: SIZES.padding,
    paddingVertical: SIZES.padding,
    flexDirection: "column",
    justifyContent: "space-evenly",
    // paddingRight: 0,
    // flex: 1,
  },
  menuText: {
    textAlign: "center",
    ...FONTS.p_regular,
    color: COLORS.accent,
  },
  menuTextAlt: {
    textAlign: "center",
    ...FONTS.p_regular,
    color: COLORS.accent,
    paddingVertical: 10,
    marginVertical: 5,
    borderColor: COLORS.accent,
    borderBottomWidth: 1,
    borderTopWidth: 1,
  },
  group: {
    margin: SIZES.margin,
    backgroundColor: COLORS.accent,
    borderRadius: SIZES.borderRadius,
    flexDirection: "row-reverse",
    overflow: "hidden",
  },
  button: {
    backgroundColor: COLORS.secondary,
    alignItems: "center",
    paddingVertical: "12%",
    paddingHorizontal: SIZES.padding,
  },
  textInput: {
    ...FONTS.h1_bold,
    color: COLORS.secondary,
    flex: 1,
    marginHorizontal: SIZES.padding,
  },
});
