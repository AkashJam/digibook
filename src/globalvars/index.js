import API from "./api";
import { UserContext, UserProvider, toastr } from "./store";

export { toastr, UserContext, UserProvider, API };
