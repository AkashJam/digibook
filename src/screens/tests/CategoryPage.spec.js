import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import renderer from "react-test-renderer";
import { UserProvider } from "../../globalvars";
import CategoryPage from "../CategoryPage";

jest.mock("react-native-root-toast", () => {
  const { View } = require("react-native");
  return {
    Toast: View,
  };
});

jest.mock("react-native-gesture-handler", () => {
  const { View } = require("react-native");
  return {
    GestureHandlerRootView: View,
    FlatList: View,
  };
});

jest.mock("@expo/vector-icons", () => {
  const { View } = require("react-native");
  return {
    MaterialCommunityIcons: View,
    Ionicons: View,
  };
});

jest.mock("react-native-maps", () => {
  const { View } = require("react-native");
  return {
    MapView: View,
    Marker: View,
  };
});

jest.mock("expo-constants", () => {
  const Constants = { isDevice: false };
  return Constants;
});

// mock navigation
const mockedNavigate = jest.fn();

jest.mock("@react-navigation/native", () => {
  return {
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe("CategoryPage", () => {
  it("renders without crashing", () => {
    const rendered = renderer
      .create(
        <UserProvider>
          <CategoryPage />
        </UserProvider>
      )
      .toJSON();
    expect(rendered).toBeTruthy();
  });

  it("control button rendered", () => {
    const { queryByText } = render(
      <UserProvider>
        <CategoryPage />
      </UserProvider>
    );
    expect(queryByText("+")).not.toBeNull();
  });

  it("dropdown menu rendered", () => {
    const { queryByText } = render(
      <UserProvider>
        <CategoryPage />
      </UserProvider>
    );
    expect(queryByText("All")).not.toBeNull();
  });

  it("card renders", () => {
    const { queryByText } = render(
      <UserProvider>
        <CategoryPage />
      </UserProvider>
    );
    expect(queryByText("test activity")).not.toBeNull();
  });

  it("card edited", () => {
    const { getByText, queryByText } = render(
      <UserProvider>
        <CategoryPage />
      </UserProvider>
    );
    fireEvent.press(getByText("test activity"));
    expect(queryByText("test activity")).not.toBeNull();
  });

  it("add card", () => {
    const msg = "Take a walk";
    const { getByText, getByPlaceholderText, queryByPlaceholderText } = render(
      <UserProvider>
        <CategoryPage />
      </UserProvider>
    );
    fireEvent.press(getByText("+"));
    fireEvent.changeText(getByPlaceholderText("Add New Task"), msg);
    fireEvent.press(getByText("+"));
    expect(queryByPlaceholderText("Add New Task")).toBeNull();
  });

  it("test against snapshot", () => {
    const tree = renderer
      .create(
        <UserProvider>
          <CategoryPage />
        </UserProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
