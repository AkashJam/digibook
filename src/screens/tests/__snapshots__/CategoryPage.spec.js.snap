// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryPage test against snapshot 1`] = `
Array [
  <div
    className="css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-u5wzg0 r-display-6koalj r-flexDirection-18u37iz r-height-1ph75f1 r-paddingHorizontal-1j3t67a r-paddingTop-sg165l r-position-u8s1d r-width-13qz1uu r-zIndex-136ojw6"
  >
    <div
      className="css-view-1dbjc4n r-cursor-1loqt21 r-touchAction-1otgn73 r-transitionProperty-1i6wzkk r-userSelect-lrvibr"
      onClick={[Function]}
      onContextMenu={[Function]}
      onKeyDown={[Function]}
      role="menu"
      style={
        Object {
          "WebkitTransitionDuration": "0s",
          "transitionDuration": "0s",
        }
      }
      tabIndex="0"
    >
      <div
        className="css-view-1dbjc4n"
      />
    </div>
    <div
      className="css-text-901oao"
      dir="auto"
      style={
        Object {
          "color": "rgba(54,54,53,1.00)",
          "fontSize": "18px",
          "marginLeft": "10px",
        }
      }
    >
      Categories
    </div>
  </div>,
  <div
    className="css-view-1dbjc4n r-backgroundColor-u5wzg0 r-height-1pi2tsx r-paddingHorizontal-1x7blby"
    style={
      Object {
        "paddingTop": "90px",
      }
    }
  >
    <div
      className="css-view-1dbjc4n r-width-13qz1uu"
      style={
        Object {
          "zIndex": 5000,
        }
      }
    >
      <div
        className="css-view-1dbjc4n r-backgroundColor-zes02y r-borderRadius-1ylenci r-borderWidth-1yadl64 r-cursor-1loqt21 r-margin-7yx52t r-paddingHorizontal-d9fdf6 r-pointerEvents-105ug2t r-touchAction-1otgn73 r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-width-1uc72sb"
        onClick={[Function]}
        onContextMenu={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "WebkitTransitionDuration": "0s",
            "transitionDuration": "0s",
            "zIndex": 5000,
          }
        }
        tabIndex="0"
      >
        <div
          className="css-text-901oao r-flex-13awgt0"
          dir="auto"
          style={
            Object {
              "color": "rgba(54,54,53,1.00)",
              "fontFamily": "Caviar_Dreams_Bold",
              "fontSize": "24px",
            }
          }
        >
          Select an item
        </div>
        <div
          className="css-view-1dbjc4n"
        >
          <div
            className="css-view-1dbjc4n r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010"
            style={
              Object {
                "height": "20px",
                "width": "20px",
              }
            }
          >
            <div
              className="css-view-1dbjc4n r-backgroundColor-1niwhzg r-backgroundPosition-vvn4in r-backgroundRepeat-u6sd8q r-backgroundSize-4gszlv r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu r-zIndex-1wyyakw"
              suppressHydrationWarning={true}
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-view-1dbjc4n r-flex-13awgt0"
      style={
        Object {
          "marginBottom": "20px",
          "marginTop": "20px",
        }
      }
    >
      <div
        className="css-view-1dbjc4n r-WebkitOverflowScrolling-150rngu r-flexDirection-eqz5dr r-flexGrow-16y2uox r-flexShrink-1wbh5a2 r-overflowX-11yh6sk r-overflowY-1rnoaur r-transform-1sncvnh"
        onScroll={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        onWheel={[Function]}
      >
        <div
          className="css-view-1dbjc4n"
        >
          <div
            className="css-view-1dbjc4n"
          >
            <div
              className="css-view-1dbjc4n"
            >
              <div
                className="css-view-1dbjc4n r-bottom-1p0dtai r-left-1d2f490 r-overflow-1udh08x r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-1h2t8mc r-zIndex-184en5c"
                style={
                  Object {
                    "height": "0px",
                  }
                }
              >
                <div
                  className="css-view-1dbjc4n r-backgroundColor-u5wzg0"
                  style={
                    Object {
                      "borderBottomColor": "rgba(0,166,251,1.00)",
                      "borderLeftColor": "rgba(0,166,251,1.00)",
                      "borderRightColor": "rgba(0,166,251,1.00)",
                      "borderTopColor": "rgba(0,166,251,1.00)",
                      "height": "0px",
                    }
                  }
                />
              </div>
              <div
                className="css-view-1dbjc4n r-zIndex-136ojw6"
                id="animatedComponent"
                onClickCapture={[Function]}
                style={
                  Object {
                    "WebkitTransform": "translateX(0px)",
                    "transform": "translateX(0px)",
                  }
                }
              >
                <div
                  className="css-view-1dbjc4n r-cursor-1loqt21 r-touchAction-1otgn73"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  tabIndex="0"
                >
                  <div
                    className="css-view-1dbjc4n r-boxShadow-hjnf4s r-elevation-1quu1zo"
                    style={
                      Object {
                        "backgroundColor": "rgba(54,54,53,1.00)",
                        "borderBottomColor": "rgba(54,54,53,1.00)",
                        "borderLeftColor": "rgba(54,54,53,1.00)",
                        "borderRightColor": "rgba(54,54,53,1.00)",
                        "borderTopColor": "rgba(54,54,53,1.00)",
                      }
                    }
                  >
                    <div
                      className="css-text-901oao"
                      dir="auto"
                      style={
                        Object {
                          "color": "rgba(246,240,237,1.00)",
                          "textDecoration": "none",
                        }
                      }
                    >
                      test activity
                    </div>
                    <div
                      className="css-view-1dbjc4n"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-view-1dbjc4n r-alignItems-1awozwy r-bottom-1p0dtai r-flexDirection-18u37iz r-justifyContent-17s6mgv r-margin-7yx52t r-padding-txj6sj r-position-u8s1d r-right-7h6s76 r-width-1h2t8mc"
    >
      <div
        className="css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-zes02y r-borderRadius-1h84pjw r-boxShadow-hjnf4s r-color-1cwaruh r-cursor-1loqt21 r-elevation-1quu1zo r-height-1ugchlj r-justifyContent-1777fci r-touchAction-1otgn73 r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-width-usgzl9"
        onClick={[Function]}
        onContextMenu={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "WebkitTransitionDuration": "0s",
            "transitionDuration": "0s",
          }
        }
        tabIndex="0"
      >
        <div
          className="css-text-901oao"
          dir="auto"
          style={
            Object {
              "fontFamily": "Caviar_Dreams_Bold",
              "fontSize": "24px",
            }
          }
        >
          +
        </div>
      </div>
    </div>
  </div>,
]
`;
